src/main/java/org/reaktivity/specification/amqp/internal/AmqpFunctions.java

```diff
@@ -192,148 +203,518 @@ public AmqpDataExBuilder flags(
+            propertiesRW.correlationId(m -> m.stringtype((String) correlationId));
             return this;
         }
 
         public AmqpDataExBuilder contentType(
             String contentType)
         {
-            dataExRW.propertiesItem(p -> p.contentType(contentType));
+            initializeProperties();
+            propertiesRW.contentType(contentType);
             return this;
         }
 
         public AmqpDataExBuilder contentEncoding(
             String contentEncoding)
         {
-            dataExRW.propertiesItem(p -> p.contentEncoding(contentEncoding));
+            initializeProperties();
+            propertiesRW.contentEncoding(contentEncoding);
             return this;
         }
 
         public AmqpDataExBuilder absoluteExpiryTime(
             long absoluteExpiryTime)
         {
-            dataExRW.propertiesItem(p -> p.absoluteExpiryTime(absoluteExpiryTime));
+            initializeProperties();
+            propertiesRW.absoluteExpiryTime(absoluteExpiryTime);
             return this;
         }
 
         public AmqpDataExBuilder creationTime(
             long creationTime)
         {
-            dataExRW.propertiesItem(p -> p.creationTime(creationTime));
+            initializeProperties();
+            propertiesRW.creationTime(creationTime);
             return this;
         }
 
         public AmqpDataExBuilder groupId(
             String groupId)
         {
-            dataExRW.propertiesItem(p -> p.groupId(groupId));
+            initializeProperties();
+            propertiesRW.groupId(groupId);
             return this;
         }
 
         public AmqpDataExBuilder groupSequence(
             int groupSequence)
         {
-            dataExRW.propertiesItem(p -> p.groupSequence(groupSequence));
+            initializeProperties();
+            propertiesRW.groupSequence(groupSequence);
             return this;
         }
 
         public AmqpDataExBuilder replyToGroupId(
             String replyToGroupId)
         {
-            dataExRW.propertiesItem(p -> p.replyToGroupId(replyToGroupId));
+            initializeProperties();
+            propertiesRW.replyToGroupId(replyToGroupId);
+            return this;
+        }
+
+        public AmqpDataExBuilder applicationProperty(
+            String key,
+            String value)
+        {
+            if (propertiesRW != null && !isPropertyBuilt)
+            {
+                final AmqpPropertiesFW properties = propertiesRW.build();
+                dataExRW.properties(properties);
+                isPropertyBuilt = true;
+            }
+            dataExRW.applicationPropertiesItem(a -> a.key(key).value(value));
             return this;
         }
 
+        private void initializeProperties()
+        {
+            if (propertiesRW == null)
+            {
+                this.propertiesRW = new AmqpPropertiesFW.Builder().wrap(new UnsafeBuffer(new byte[1024]), 0, 1024);
+            }
+        }
```

Comment:
Suggest renaming this to `properties` and returning `propertiesRW`.
Then usages can call `properties().XXX` instead of a separate initialize call relying on side effect before accessing `propertiesRW`.